libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1239 @@
+{
+  g_autoptr(ArtworkCardsFromShardsAndItemsData) marshal_data =
+    artwork_cards_from_shards_and_items_data_new (ka_proxy,
+                                                  EOS_DISCOVERY_FEED_CARD_LAYOUT_DIRECTION_IMAGE_LAST,
+                                                  EOS_DISCOVERY_FEED_CARD_STORE_TYPE_ARTWORK_CARD,
+                                                  EOS_DISCOVERY_FEED_THUMBNAIL_SIZE_ARTWORK);
+  GDBusProxy *dbus_proxy = eos_discovery_feed_knowledge_app_proxy_get_dbus_proxy (ka_proxy);
+  const gchar *desktop_id = eos_discovery_feed_knowledge_app_proxy_get_desktop_id (ka_proxy);
+
+  return call_dbus_proxy_and_construct_from_models_and_shards (dbus_proxy,
+                                                               "ArtworkCardDescriptions",
+                                                               desktop_id,
+                                                               artwork_cards_from_shards_and_items,
+                                                               marshal_data,
+                                                               cancellable,
+                                                               error);
+}
+
+typedef gpointer (*AppendStoresFromProxyFunc) (EosDiscoveryFeedKnowledgeAppProxy  *ka_proxy,
+                                               gpointer                            proxy_data,
+                                               GCancellable                       *cancellable,
+                                               GError                            **error);
+
+typedef struct _AppendStoresTaskData
+{
+  EosDiscoveryFeedKnowledgeAppProxy *ka_proxy;
+  AppendStoresFromProxyFunc          proxy_func;
+  gpointer                           proxy_data;
+} AppendStoresTaskData;
+
+static AppendStoresTaskData *
+append_stores_task_data_new (EosDiscoveryFeedKnowledgeAppProxy *ka_proxy,
+                             AppendStoresFromProxyFunc          proxy_func,
+                             gpointer                           proxy_data)
+{
+  AppendStoresTaskData *data = g_new0 (AppendStoresTaskData, 1);
+
+  data->ka_proxy = g_object_ref (ka_proxy);
+  data->proxy_func = proxy_func;
+  data->proxy_data = proxy_data;
+
+  return data;
+}
+
+static void
+append_stores_task_data_free (AppendStoresTaskData *data)
+{
+  g_clear_object (&data->ka_proxy);
+
+  g_free (data);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (AppendStoresTaskData,
+                               append_stores_task_data_free);
+
+static void
+append_stores_task_from_proxy_thread (GTask        *task,
+                                      gpointer      source,
+                                      gpointer      task_data,
+                                      GCancellable *cancellable)
+{
+  AppendStoresTaskData *data = task_data;
+  g_autoptr(GError) local_error = NULL;
+
+  g_autoptr(GPtrArray) results = data->proxy_func (data->ka_proxy,
```

Comment:
Ouch okay, this does not make much sense - `data->proxy_func` can return anything, we should be passing our own `GDestroyNotify` here.